testcases/OpalGard.py

```diff
@@ -0,0 +1,103 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing
+# permissions and limitations under the License.
+#
+# IBM_PROLOG_END_TAG
+
+# @package opal-gard
+#  Test different OPAL GARD Related functionality
+#
+
+import re
+import random
+
+
+from common.OpTestConstants import OpTestConstants as BMC_CONST
+import unittest
+
+import OpTestConfiguration
+from common.OpTestSystem import OpSystemState
+from common.Exceptions import CommandFailed
+
+class OpalGard(unittest.TestCase):
+    def setUp(self):
+        conf = OpTestConfiguration.conf
+        self.cv_IPMI = conf.ipmi()
+        self.cv_SYSTEM = conf.system()
+        self.cv_HOST = conf.host()
+        self.bmc_type = conf.args.bmc_type
+        self.cv_SYSTEM.goto_state(OpSystemState.OS)
+
+
+    def list_gard_records(self):
+        cmd = "PATH=/usr/local/sbin:$PATH opal-gard list all"
+        try:
+            res = self.c.run_command(cmd,timeout=120)
+        except CommandFailed as cf:
+            print str(cf)
+            self.gather_debug_logs()
+            raise Exception("List gard records operation failed")
+
+    def clear_gard_records(self):
+        cmd = "PATH=/usr/local/sbin:$PATH opal-gard clear all"
+        try:
+            res = self.c.run_command(cmd,timeout=120)
+        except CommandFailed as cf:
+            print str(cf)
+            self.gather_debug_logs()
+            raise Exception("Clear gard records operation failed")
+
+    def show_gard_record(self, id):
+        cmd = "PATH=/usr/local/sbin:$PATH opal-gard show %s" % id
+        try:
+            res = self.c.run_command(cmd,timeout=120)
+        except CommandFailed as cf:
+            print str(cf)
+            self.gather_debug_logs()
+            raise Exception("show gard records operation failed")
+
+    def gather_debug_logs(self):
+        cmd = "dmesg -T --level=alert,crit,err,warn"
+        res = self.c.run_command(cmd, timeout=120)
+        self.c.run_command_ignore_fail("grep ',[0-4]\]' /sys/firmware/opal/msglog")
+
+    def runTest(self):
+        # opal-gard from host is not supported in FSP systems
+        if "FSP" in self.bmc_type:
+            self.skipTest("OP BMC specific")
+
+        self.c = self.cv_SYSTEM.sys_get_ipmi_console()
+        self.cv_SYSTEM.host_console_login()
+	self.cv_SYSTEM.host_console_unique_prompt()
+        self.cv_HOST.host_copy_fake_gard()
+        self.c.run_command("dmesg -D")
+        self.cpu = ''.join(self.c.run_command("grep '^cpu' /proc/cpuinfo |uniq|sed -e 's/^.*: //;s/ .*//;'"))
+
+        if self.cpu not in ["POWER8", "POWER8E", "POWER9"]:
+            self.skipTest("Unknown CPU type %s" % self.cpu)
+        for i in range(0, 11):
+            self.list_gard_records()
+            self.c.run_command("dd if=/tmp/fake.gard of=/dev/mtd0 bs=$((0x10)) seek=$((0x2c00)) conv=notrunc")
```

Comment:
If ECC is enabled then directly writing data like above may create issue during next boot where it tries to read gard records. You may want to dig more on this.


-Vasant

Reply:
I have tried to isolate the issue https://github.com/open-power/boston-openpower/issues/568. based on our Test.

1. Just flash fake.gard using above dd command and reboot the system. this test works fine for multiple times. and system boots fine without any PNOR corruption.

2. But when we include GARD API after flashing, at some point of time opal-gard command hangs and immediate reboot causes the boot failure.

So i guess, the above command has no problems with PNOR corruption.

Reply:
The host still boots after writing the record, so I can only assume that fake record has the correct ECC information.

Reply:
I *really* don't want us to hardcode the location of GUARD partition though. Could we instead just add pflash to the list of things you need to have installed on the host and use it instead?